tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,367 @@
+import pandas as pd
+import numpy as np
+
+from tardis.plasma.properties.base import ProcessingPlasmaProperty
+
+__all__ = [
+    "NLTERateEquationSolver",
+]
+
+
+class NLTERateEquationSolver(ProcessingPlasmaProperty):
+    outputs = ("ion_number_density_nlte", "electron_densities_nlte")
+
+    def calculate(
+        self,
+        gamma,
+        alpha_sp,
+        alpha_stim,
+        coll_ion_coeff,
+        coll_recomb_coeff,
+        partition_function,
+        levels,
+        level_boltzmann_factor,
+        phi,
+        rate_matrix_index,
+        number_density,
+    ):
+        """Calculates ion number densities and electron densities using NLTE ionization.
+
+        Parameters
+        ----------
+        gamma : DataFrame
+        alpha_sp : DataFrame
+        alpha_stim : DataFrame
+        coll_ion_coeff : DataFrame
+        coll_recomb_coeff : DataFrame
+        partition_function : DataFrame
+        levels : MultiIndex
+        level_boltzmann_factor : DataFrame
+        phi : DataFrame
+        rate_matrix_index : MultiIndex
+        number_density : DataFrame
+
+        Returns
+        -------
+        ion_number_densities_nlte: DataFrame
+        electron_densities_nlte: Series
+        """
+
+        (
+            total_photo_ion_coefficients,
+            total_rad_recomb_coefficients,
+            total_coll_ion_coefficients,
+            total_coll_recomb_coefficients,
+        ) = self.prepare_ion_recomb_coefficients_nlte_ion(
+            gamma,
+            alpha_sp,
+            alpha_stim,
+            coll_ion_coeff,
+            coll_recomb_coeff,
+            partition_function,
+            levels,
+            level_boltzmann_factor,
+        )
+
+        # >>>TODO:initial electron density should be included in the initial guess, added in a future PR
+        initial_electron_density = number_density.sum(axis=0)
+        # <<<
+        rate_matrix = self.calculate_rate_matrix(
+            phi[0],
+            initial_electron_density[0],
+            rate_matrix_index,
+            total_photo_ion_coefficients[0],
+            total_rad_recomb_coefficients[0],
+            total_coll_ion_coefficients[0],
+            total_coll_recomb_coefficients[0],
+        )
+        return (
+            -1,
```

Comment:
`None` is probably a better return than -1

Reply:
Either that or `raise NotImplementedError`

Reply:
Done